### DISCUSSION TITLE
Custom domain & soft deploy

### DISCUSSION THREAD

**Discussion by dmudro** [vercel/community#745] [Idea]
occasionally I'd love to have the ability to build & deploy the production branch but without prod custom domain mapping.

heavy features branches would benefit from *soft deploy* to underlying build domain first (i. e. your *.vercel.app) so we can test sensitive areas before mapping the build to the production URL.

Vercel previews aka dev branches are great though these environments are typically mapped to dev database, dev CMS etc. such setup is mostly fine for QA but in some areas you dont really know until going to produciton and deal with real data. 

the point here is having the ability to pre-relaase to prod "internally", i. e. Next.js app mapped to other prod envs for internal testing and against all production services.

**Comment by dmudro** [comment]
the idea occurred to me when we had to rollback a failing release. so we did and after the rollback I was still able to debug the failing build in production - Vercel deployments keep all builds on non public build domains. I was able to debug the production issue on this domain while our public custom domain kept serving the rolled back version. 

because of our dev services setup and the nature of quality of dev data, we had not been able to catch the issue in preview. I trust this is not uncommon and while it does not happen often it's usually the production-only-sensitive-hard-to-debug-issues.

**Comment by amyegan** [comment]
Thanks for sharing this idea! You can accomplish something similar at the moment by setting up a [staging environment](https://vercel.com/guides/set-up-a-staging-environment-on-vercel) with a specific branch. You can use a dedicated custom domain or subdomain with the staging environment, which is much easier to remember than the generated URL for individual deployments. You could use this method to make a "production-test" environment for testing and debugging those difficult problems without disrupting your production site.

**Reply by dmudro** [reply]
thanks for the feedback. I already use a "staging" environment with a custom domain. but since there is only one preview env in Vercel, it is mapped to our staging services.

additionally your suggestion wouldnt really work with the github flow aka feature branch flow - which Vercel CI is pretty much based on.

**Comment by karlomedallo** [comment]
We are using Vercel CLI to deploy since we use a self-hosted gitlab instance and does not have access to "per branch specific" variables.
This idea is what we wanted so that we can have a production deployment `vercel deploy --prod` but not having the custom domains assigned yet.
I would think that a simple switch in vercel cli can do this. Maybe something like `vercel deploy --prod --skip-custom-domain`
The expectation is that the normal production deployment happens but only Vercel generated URLs are assigned to it.

By doing this, we can test that deployment first via Vercel generated URLs. We can then just use `vercel alias set` to this new deployment to assign the custom domain and make it the "Current" production deployment.

**Comment by cpf-sh** [comment]
The lack of this functionality is a real pain point when trying to orchestrate deployments across multiple different providers with a tightly-coupled legacy application. It's a great shame it's not supported by Vercel.
